pgjdbc/src/main/java/com/yugabyte/ysql/LoadBalanceManager.java

```diff
@@ -0,0 +1,459 @@
+        list.add(h);
+      }
+    }
+    return list;
+  }
+
+  public static int getPort(String host) {
+    NodeInfo info = clusterInfoMap.get(host);
+    return info != null ? info.port : 5433;
+  }
+
+  public static boolean incrementConnectionCount(String host) {
+    NodeInfo info = clusterInfoMap.get(host);
+    if (info != null) {
+      synchronized (info) {
+        if (info.connectionCount < 0) {
+          info.connectionCount = 0;
+          LOGGER.fine("Resetting connection count for " + host + " to zero from " + info.connectionCount);
+        }
+        info.connectionCount += 1;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public static boolean decrementConnectionCount(String host) {
+    NodeInfo info = clusterInfoMap.get(host);
+    if (info != null) {
+      synchronized (info) {
+        info.connectionCount -= 1;
+        LOGGER.fine("Decremented connection count for " + host + " by one: " + info.connectionCount);
+        if (info.connectionCount < 0) {
+          info.connectionCount = 0;
+          LOGGER.fine("Resetting connection count for " + host + " to zero.");
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public static Connection getConnection(String url, Properties properties, String user,
+      String database) {
+    LoadBalanceProperties lbProperties = LoadBalanceProperties.getLoadBalanceProperties(url,
+        properties);
+    // Cleanup extra properties used for load balancing?
+    if (lbProperties.hasLoadBalance()) {
+      Connection conn = getConnection(lbProperties,  user, database);
+      if (conn != null) {
+        return conn;
+      }
+      LOGGER.warning("Failed to apply load balance. Trying normal connection");
+    }
+    return null;
+  }
+
+  public static Connection getConnection(LoadBalanceProperties loadBalanceProperties, String user,
+      String dbName) {
+    LoadBalancer lb = loadBalanceProperties.getAppropriateLoadBalancer();
+    Properties props = loadBalanceProperties.getOriginalProperties();
+    String url = loadBalanceProperties.getStrippedURL();
+
+    if (!checkAndRefresh(loadBalanceProperties, lb, user, dbName)) {
+      return null;
+    }
+
+    List<String> failedHosts = new ArrayList<>();
+    String chosenHost = lb.getLeastLoadedServer(true, failedHosts);
+    PgConnection newConnection;
+    SQLException firstException = null;
+    while (chosenHost != null) {
+      try {
+        props.setProperty("PGHOST", chosenHost);
+        props.setProperty("PGPORT", String.valueOf(LoadBalanceManager.getPort(chosenHost)));
+        newConnection = new PgConnection(hostSpecs(props), user, dbName, props, url);
+        newConnection.setLoadBalancer(lb);
+        LOGGER.fine("Created connection to " + chosenHost);
+        return newConnection;
+      } catch (SQLException ex) {
+        // Let the refresh be forced the next time it is tried. todo Is this needed?
+        forceRefreshOnce = true;
```

Comment:
Is this consistent across all smart drivers? I could not locate this logic in `pgx smart driver`.